packages/js/data/src/crud/reducer.ts

```diff
@@ -101,9 +101,47 @@ export const createReducer = (
 					};
 
 					let items = state.items;
+					let queryItems = Object.keys( data ).map( ( key ) => +key );
+
 					let itemsCount = state.itemsCount;
 
-					if ( typeof options?.optimisticQueryUpdate === 'object' ) {
+					/*
+					 * Check it needs to update the store with the new item,
+					 * optimistically.
+					 */
+					if ( options?.optimisticQueryUpdate ) {
+						/*
+						 * If the query has an order_by property, sort the items
+						 * by the order_by property.
+						 *
+						 * The sort criteria could be different from the
+						 * the server side.
+						 * Ensure to keep in sync with the server side, for instance,
+						 * by invalidating the cache.
+						 *
+						 * Todo: Add a mechanism to use the server side sorting criteria.
+						 */
+						if ( options.optimisticQueryUpdate?.order_by ) {
+							type OrderBy = keyof Item;
+							const order_by = options.optimisticQueryUpdate
+								?.order_by as OrderBy;
+
+							let sortingData = Object.values( data );
+							sortingData = sortingData.sort( ( a, b ) =>
+								( a[ order_by ] as string )
+									.toLowerCase()
+									.localeCompare(
+										(
+											b[ order_by ] as string
+										 ).toLowerCase()
+									)
+							);
```

Comment:
Something like this is ok for now, although this may get complicated quickly if there were any other query params supported.
We should keep the server as source of truth.
I think it would be fine to trigger a refresh in-behind the scenes without showing it to the user.

Reply:
100% agree. I added an important comment about it